# Project Overview

## Import libraries and load dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.options.mode.chained_assignment = None
pd.set_option('future.no_silent_downcasting', True)

In [3]:
data = pd.read_csv(r'D:\REGINA\yesgo\0 NEW\tata\Online Retail Data Set.csv')
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## Data cleaning and preparation

In [4]:
# Filter data with quantity less than 1
data[data['Quantity'] < 1]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

There are 10,624 rows where the quantity value is 0 or negative. We can exclude these rows from the analysis.

In [5]:
# Remove rows with quantity less than 1
data = data[~(data['Quantity'] < 1)]

In [6]:
# Verify rows with quantity less than 1
data[data['Quantity'] < 1]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [7]:
# Filter data with unit price less than 0
data[data['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

There are 2 rows where the unit price value is negative. We can exclude these rows from the analysis.

In [8]:
# Remove rows with unit price less than 0
data = data[~(data['UnitPrice'] < 0)]

In [9]:
# Verify rows with unit price less than 0
data[data['UnitPrice'] < 0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 531283 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    531283 non-null  object 
 1   StockCode    531283 non-null  object 
 2   Description  530691 non-null  object 
 3   Quantity     531283 non-null  int64  
 4   InvoiceDate  531283 non-null  object 
 5   UnitPrice    531283 non-null  float64
 6   CustomerID   397924 non-null  float64
 7   Country      531283 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 36.5+ MB


In [11]:
# Convert the invoice date data type to datetime for analysis
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

In [12]:
# Calculate the percentage of duplicate rows
data.duplicated().sum() / len(data) *100

0.9845976626393089

In [13]:
# Check duplicated rows
data[data.duplicated(keep=False)].sort_values(by=['InvoiceNo', 'StockCode'])

InvoiceNo StockCode                       Description  Quantity  \
494       536409     21866       UNION JACK FLAG LUGGAGE TAG         1   
517       536409     21866       UNION JACK FLAG LUGGAGE TAG         1   
485       536409     22111      SCOTTIE DOG HOT WATER BOTTLE         1   
539       536409     22111      SCOTTIE DOG HOT WATER BOTTLE         1   
489       536409     22866     HAND WARMER SCOTTY DOG DESIGN         1   
...          ...       ...                               ...       ...   
541655    581538     23275  SET OF 3 HANGING OWLS OLLIE BEAK         1   
541656    581538     23318    BOX OF 6 MINI VINTAGE CRACKERS         1   
541689    581538     23318    BOX OF 6 MINI VINTAGE CRACKERS         1   
541666    581538     23343      JUMBO BAG VINTAGE CHRISTMAS          1   
541701    581538     23343      JUMBO BAG VINTAGE CHRISTMAS          1   

               InvoiceDate  UnitPrice  CustomerID         Country  
494    2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
517    2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
485    2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
539    2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
489    2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
...                    ...        ...         ...             ...  
541655 2011-12-09 11:34:00       1.25     14446.0  United Kingdom  
541656 2011-12-09 11:34:00       2.49     14446.0  United Kingdom  
541689 2011-12-09 11:34:00       2.49     14446.0  United Kingdom  
541666 2011-12-09 11:34:00       2.08     14446.0  United Kingdom  
541701 2011-12-09 11:34:00       2.08     14446.0  United Kingdom  

[10078 rows x 8 columns]

A total of 5,231 rows are duplicated and can be removed.

In [14]:
# Delete duplicate rows
data = data.drop_duplicates()

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 526052 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    526052 non-null  object        
 1   StockCode    526052 non-null  object        
 2   Description  525460 non-null  object        
 3   Quantity     526052 non-null  int64         
 4   InvoiceDate  526052 non-null  datetime64[ns]
 5   UnitPrice    526052 non-null  float64       
 6   CustomerID   392732 non-null  float64       
 7   Country      526052 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 36.1+ MB


In [16]:
# Find null rows in the description column
data[data['Description'].isna()]

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
622       536414     22139         NaN        56 2010-12-01 11:52:00   
1970      536545     21134         NaN         1 2010-12-01 14:32:00   
1971      536546     22145         NaN         1 2010-12-01 14:33:00   
1972      536547     37509         NaN         1 2010-12-01 14:33:00   
1987      536549    85226A         NaN         1 2010-12-01 14:34:00   
...          ...       ...         ...       ...                 ...   
533712    581103     22689         NaN         4 2011-12-07 11:58:00   
535326    581203     23406         NaN        15 2011-12-07 18:31:00   
535332    581209     21620         NaN         6 2011-12-07 18:35:00   
536981    581234     72817         NaN        27 2011-12-08 10:33:00   
538554    581408     85175         NaN        20 2011-12-08 14:06:00   

        UnitPrice  CustomerID         Country  
622           0.0         NaN  United Kingdom  
1970          0.0         NaN  United Kingdom  
1971          0.0         NaN  United Kingdom  
1972          0.0         NaN  United Kingdom  
1987          0.0         NaN  United Kingdom  
...           ...         ...             ...  
533712        0.0         NaN  United Kingdom  
535326        0.0         NaN  United Kingdom  
535332        0.0         NaN  United Kingdom  
536981        0.0         NaN  United Kingdom  
538554        0.0         NaN  United Kingdom  

[592 rows x 8 columns]

In [17]:
# Fill null values in the description column using stock code, and 'No description' for others.
data['Description'] = data.groupby('StockCode')['Description'].transform(lambda x: x.ffill().bfill())
data['Description'] = data['Description'].fillna('No Description')

In [18]:
# Fill missing values in the customer ID column
data['CustomerID'] = data['CustomerID'].apply(lambda x: str(int(x)) if not pd.isna(x) else 'No Customer ID')

In [19]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 526052 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    526052 non-null  object        
 1   StockCode    526052 non-null  object        
 2   Description  526052 non-null  object        
 3   Quantity     526052 non-null  int64         
 4   InvoiceDate  526052 non-null  datetime64[ns]
 5   UnitPrice    526052 non-null  float64       
 6   CustomerID   526052 non-null  object        
 7   Country      526052 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 36.1+ MB


The dataset is cleaned and prepared for analysis.

In [20]:
data.to_csv(r'D:\REGINA\yesgo\0 NEW\tata\clean_data.csv', index=False)